server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java

```diff
@@ -978,20 +988,74 @@ public long countCpusForAccount(long accountId) {
-            sc1.setParameters("state", new Object[] {State.Destroyed, State.Error, State.Expunging, State.Stopped});
-        else
-            sc1.setParameters("state", new Object[] {State.Destroyed, State.Error, State.Expunging});
+        if (VirtualMachineManager.ResourceCountRunningVMsonly.value()) {
+            sc1.setParameters("state", State.Destroyed, State.Error, State.Expunging, State.Stopped);
+        } else {
+            sc1.setParameters("state", State.Destroyed, State.Error, State.Expunging);
+        }
         sc1.setParameters("displayVm", 1);
         List<UserVmJoinVO> userVms = _userVmJoinDao.search(sc1,null);
         for (UserVmJoinVO vm : userVms) {
-            cputotal += Long.valueOf(vm.getCpu());
+            cputotal += vm.getCpu();
+        }
+
+        final Pair<Boolean, Long> totalByAccountId = calculateCpuTotalByAccountId(accountId);
+        if (totalByAccountId.first()) {
+            return totalByAccountId.second();
+        } else {
+            return cputotal;
         }
-        return cputotal;
+    }
+
+    private Pair<Boolean, Long> calculateCpuTotalByAccountId(long accountId) {
+        long cputotal = 0;
+        boolean hasCpus = false;
+        final Account owner = _accountDao.findById(accountId);
+        final DomainVO domain = _domainDao.findById(owner.getDomainId());
+        final ServiceOffering defaultRouterOffering = getServiceOfferingByConfig();
+
+        GenericSearchBuilder<ServiceOfferingVO, SumCount> cpuSearch = serviceOfferingDao.createSearchBuilder(SumCount.class);
+        cpuSearch.select("sum", Func.SUM, cpuSearch.entity().getCpu());
+        cpuSearch.select("count", Func.COUNT, (Object[])null);
+        SearchBuilder<VMInstanceVO> join1 = _vmDao.createSearchBuilder();
+        join1.and("accountId", join1.entity().getAccountId(), Op.EQ);
+        join1.and("type", join1.entity().getType(), Op.IN);
+        join1.and("state", join1.entity().getState(), SearchCriteria.Op.NIN);
+        join1.and("displayVm", join1.entity().isDisplayVm(), Op.EQ);
+        cpuSearch.join(OFFERINGS_NAME, join1, cpuSearch.entity().getId(), join1.entity().getServiceOfferingId(), JoinBuilder.JoinType.INNER);
+        cpuSearch.done();
+
+        SearchCriteria<SumCount> sc = cpuSearch.create();
+        sc.setJoinParameters(OFFERINGS_NAME, "accountId", accountId);
+        sc.setJoinParameters(OFFERINGS_NAME, "type", VirtualMachine.Type.DomainRouter); // domain routers
+
+        if (Boolean.TRUE.equals(VirtualMachineManager.ResourceCountRunningVMsonly.value())) {
+            sc.setJoinParameters(OFFERINGS_NAME, "state", State.Destroyed, State.Error, State.Expunging, State.Stopped);
+        } else {
+            sc.setJoinParameters(OFFERINGS_NAME, "state", State.Destroyed, State.Error, State.Expunging);
+        }
+        sc.setJoinParameters(OFFERINGS_NAME, "displayVm", 1);
+        List<SumCount> cpus = serviceOfferingDao.customSearch(sc, null);
```

Comment:
I would refactor this to a method in the serviceOfferingDao, this is mostly duplicated code from another method.